### PR TITLE
retry domain boot on shutdown from inside

### PR DESCRIPTION
**Pull request by giggsoff**
In case of HALTED state (and Activate state from controller) we should retry boot domain as discussed in https://github.com/lf-edge/eve/pull/2305#pullrequestreview-772878297

Signed-off-by: Petr Fedchenkov <giggsoff@gmail.com>

### CODE REVIEW

File: pkg/pillar/cmd/domainmgr/domainmgr.go [hunk 1]
```diff
@@ -849,13 +852,18 @@ func verifyStatus(ctx *domainContext, status *types.DomainStatus) {
 				err := fmt.Errorf("one of the %s tasks has crashed (%v)", status.Key(), err)
 				log.Errorf(err.Error())
 				status.SetErrorNow("one of the application's tasks has crashed - please restart application instance")
```
Comment (eriknordmark) [inline pkg/pillar/cmd/domainmgr/domainmgr.go:854]: Do we ever hit this broken case when the app instance is shutdown or halted from inside?

**Comment by eriknordmark** [comment]
@adarsh-zededa can you take a look?

**Comment by adarsh-zededa** [comment]
Had a separate chat with @petr-zededa regarding this. This solution in the PR might cause an issue as the domain goes to HALTED state even during a crash and in such cases, it's best to leave it as it is to debug instead of restarting it. 
The correct solution would be to add a check in [this method](https://github.com/lf-edge/eve/blob/d2fc8cdc5f289668ee0bb71c2c1885e34ebe1cde/pkg/pillar/hypervisor/qmp.go#L86) and handle manual shutdown separately. Now when I was trying that out, I noticed for some cases we was not reliably getting the `SHUTDOWN` state from qmp process. I was checking what other states can we monitor (list of states can be found [here](https://github.com/coreos/qemu/blob/master/docs/qmp-events.txt))

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 1 above)
Comment (giggsoff) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:854]: I assume that for broken state we want only notify the user (as described in the comment). My attempts to shutdown domain from inside leads to the halted state.
Halt from inside do not lead to any event at all and just hung the VM in running state, but it looks like the separate issue.

**Comment by giggsoff** [comment]
> Had a separate chat with @petr-zededa regarding this. This solution in the PR might cause an issue as the domain goes to `HALTED` state even during a crash and in such cases, it's best to leave it as it is to debug instead of restarting it.

In this case we still will have information of state transition of VM. In general for this retrial logic we want to try to start until we receive activate=false from controller. So, if we want to debug any issues, we can manually stop it and go through logs.

**Comment by giggsoff** [comment]
@adarsh-zededa do you have any comments on this? I still cannot found what to improve in `qmpEventHandler`. In case of halted domain (no qemu process) and activate state from config seems reasonable to try to boot it again.

**Comment by adarsh-zededa** [comment]
> @adarsh-zededa do you have any comments on this? I still cannot found what to improve in `qmpEventHandler`. In case of halted domain (no qemu process) and activate state from config seems reasonable to try to boot it again.

@petr-zededa I found the same in my experiments as well where deciding whether an app crashed or was powered down based on qmp events was inconclusive as we got `SHUTDOWN` in both the cases (since we called `/sbin/poweroff` in both the cases [here](https://github.com/lf-edge/eve/blob/4e28e007225501654b77fd8d27ebe0e7f06b61ff/pkg/xen-tools/initrd/init-initrd#L113)). So the solution you proposed makes sense where we keep retrying until the domain's activate flag is set. But, IMHO we might have to get CS team's opinion as well on whether they agree with this behaviour.

**Comment by eriknordmark** [comment]
> @petr-zededa I found the same in my experiments as well where deciding whether an app crashed or was powered down based on qmp events was inconclusive as we got `SHUTDOWN` in both the cases (since we called `/sbin/poweroff` in both the cases [here](https://github.com/lf-edge/eve/blob/4e28e007225501654b77fd8d27ebe0e7f06b61ff/pkg/xen-tools/initrd/init-initrd#L113)). So the solution you proposed makes sense where we keep retrying until the domain's activate flag is set. But, IMHO we might have to get CS team's opinion as well on whether they agree with this behaviour.

Another consideration is how quickly it will retry if the app instance fails to boot for some unknow reason. Do we always wait for the boot retry timer? Or are there cases where we can quickly retry causing lots of unneeded load if the app instance crashes each time on boot? Would it make sense to look at the uptime of the app instance and if that is more than say 5 minutes we immediately restart it, otherwise we wait for the boot retry timer?

**Comment by giggsoff** [comment]
Modified the code to indicate an error in status and in turn rely on [domainBootRetryTime](https://github.com/lf-edge/eve/blob/ef4a9efd9b3b05a4e8b243928bddee8c1f3bd252/pkg/pillar/cmd/domainmgr/domainmgr.go#L930) to not repeat immediately,

**Review by eriknordmark — APPROVED** [review]
Kick off tests

**Review by rvs — COMMENTED** [review]
LGTM

**Pull request merged**